DT_FLASH_AREA replacement with new dts macros

galak opened the pull request.

galak committed:
devicetree: Add support for fixed-partitions

Add DT_NODE_BY_FIXED_PARTITION_LABEL that given a "label" in any
fixed-partitions map will return the node_id for that partition node.

Add DT_NODE_HAS_FIXED_PARTITION_LABEL that will test if a given
fixed-partitions "label" is valid.

Add DT_FIXED_PARTITION_ID that will return an unique ordinal value for
the partition give a node_id to the partition.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
edtlib: handle match of fixed-partition on any bus

If we have a fixed-partition on a flash device that is for example on
a spi controller we will not get a binding match currently.  This is
because we expect a match between both the compatible and the fact that
fixed-partition node is a decendant of the spi bus.

To address this we treat fixed-partitions as if they are on no bus.
This has the effect of causing a binding match as well as ensuring that
when we process the fixed-partition node we will do anything special to
it because of the bus it happens to be under (for example SPI CS_GPIO
processing).

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
storage: flash_map_default: Convert to new dts macros

Convert to use new DTS macros to populate default flash map

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
storage: flash_map: Add macros to get FLASH_AREA info

Add macros that we'll utilize instead of DT_FLASH_AREA_

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
flash: Convert DT_FLASH_AREA to FLASH_AREA macros

Convert with a combo of scripts and by hand fixups:

git grep -l DT_FLASH_AREA_.*_ID | \
 xargs sed -i -r 's/DT_FLASH_AREA_(.*)_ID/FLASH_AREA_ID(\L\1)/'

git grep -l DT_FLASH_AREA_.*_OFFSET | \
 xargs sed -i -r 's/DT_FLASH_AREA_(.*)_OFFSET/FLASH_AREA_OFFSET(\L\1)/'

git grep -l DT_FLASH_AREA_.*_SIZE | \
 xargs sed -i -r 's/DT_FLASH_AREA_(.*)_SIZE/FLASH_AREA_SIZE(\L\1)/'

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak commented:
This is based off of https://github.com/zephyrproject-rtos/zephyr/pull/24522.

There's a known issue with samples/subsys/fs/littlefs on nrf52840dk_nrf52840 failing.  This is because the nrf52840dk_nrf52840 has 2 fixed-partition nodes.  We need to decide if partitions should be seen as per flash device or more global.

pabigot commented:
> There's a known issue with samples/subsys/fs/littlefs on nrf52840dk_nrf52840 failing.

I wasn't aware of that issue.  That's a very ugly failure message.

> This is because the nrf52840dk_nrf52840 has 2 fixed-partition nodes. We need to decide if partitions should be seen as per flash device or more global.

Aha.  Now I finally see what may be underlying the discussion at https://github.com/zephyrproject-rtos/zephyr/pull/24522#discussion_r416022811

Flash partitions identified by `label` property should be global; i.e. the name should be sufficient to identify the containing device and region within the device.  If the numeric ID is local to the containing device it shouldn't matter.

de-nordic commented:
> 
> 
> > There's a known issue with samples/subsys/fs/littlefs on nrf52840dk_nrf52840 failing.

I have seen that in original PR for FLASH_AREA. I have posted workaround PR here:
https://github.com/zephyrproject-rtos/zephyr/pull/25062

galak committed:
edtlib: Initialize node.compats earlier

Setup node.compats right after we create the Node.  This allows access
to the compats information in _bus_node.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
Kconfig: Default to 'n' for LEGACY_DEVICETREE_MACROS

Set LEGACY_DEVICETREE_MACROS to default to no since we are deprecating
the old macro style and all in tree users are now converted.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
west.yml: Update mcumgr/mcuboot to pickup changes to match new dts

Update mcumgr/mcuboot for changes to DT_FLASH_AREA -> FLASH_AREA macros.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak commented:
This should be fully functional now.

galak commented:
@de-nordic this should supersede #25062, #24848, #24644, and #24514.  We should pull the tests out of #25062. (but not wanting to add more to this PR to try to get it to closure).

mbolivar-nordic commented on scripts/dts/gen_defines.py:
```diff
@@ -360,6 +366,23 @@ def write_compatibles(node):
+    # Writes macro that are helpers that will call a macro/function
+    # for each child node.
+
+    if node.children:
+        functions = ''
+        # For each subnode with non-empty list of labels add the
+        # first (non-alias) label to the string of ', ' separated
+        # list of labels.
+        for subnode in node.children.values():
+            functions = functions + f"fn(DT_{subnode.z_path_id}) "
+
+        if functions:
+            macro = f"{node.z_path_id}_FOREACH_CHILD(fn)"
+            out_dt_define(macro, functions)
```
```suggestion    if node.children:
        out_dt_define(f"{node.z_path_id}_FOREACH_CHILD(fn)",
                      " ".join(f"fn(DT_{child.z_path_id})" for child in
                               node.children.values()))

```

mbolivar-nordic commented on scripts/dts/gen_defines.py:
```diff
@@ -252,6 +252,10 @@ def write_special_props(node):
     write_compatibles(node)
 
+    if "fixed-partitions" in node.parent.compats:
+        macro = f"{node.z_path_id}_PARTITION_ID"
```
Why not just emit `{node.z_path_id}_DEP_ORDINAL` for every node? Less of a special case that way.

galak replied:
I was being conservative at this point.  I'm good w/renaming the macro to `_DEP_ORDINAL`.  I'd like to limit it to `fixed-partitions` for now until we figure out the next phase of DT w/regards to 'struct device'

mbolivar-nordic replied:
Okay. Reasonable.

mbolivar-nordic commented on scripts/dts/gen_defines.py:
```diff
@@ -575,6 +579,18 @@ def write_global_compat_info(edt):
+                for child in node.children.values():
+                    if "label" in child.props:
+                        label = child.props["label"].val
+                        macro = f"COMPAT_{str2ident(compat)}_LABEL_{str2ident(label)}"
```
Is this meant as a good-enough-for-now solution? I am asking because I'm not sure labels are unique in general. I know non-unique labels in zephyr devicetrees break device_get_binding(), so we rely on unique labels, but I'm asking since we are getting away from requiring label properties in general.

Is it possible to use a node label like `boot_partition` instead of a label property like `mcuboot` here? Node labels are unique, after all.

mbolivar-nordic replied:
I think I have answered my own question. I am OK with labels since dtspec says the value of the property is binding-specific and upstream makes it seem like they're expected to seem unique, as pointed out by @pabigot earlier.

galak replied:
Its meant as good-enough and maintaining the functionality we have today.  I'd prefer to stay away from nodelabel's here especially with the long standing use of the `label` property.

I'd be interested to see how we progress the flash map API w/regards to multiple flash's.

We can also add something to the binding (upstream) about label being unique.

mbolivar-nordic commented on Kconfig.zephyr:
```diff
@@ -350,7 +350,7 @@ config MAKEFILE_EXPORTS
 config LEGACY_DEVICETREE_MACROS
 	bool "Allow use of legacy devicetree macros"
-	default y
+	default n
```
Can just remove the line. This will probably anger the linter.

mbolivar-nordic replied:
Doesn't this break mcuboot?

```
=== running "rg DT_FLASH_AREA" in mcuboot (bootloader/mcuboot):
scripts/assemble.py
37:offset_re = re.compile(r"^#define DT_FLASH_AREA_([0-9A-Z_]+)_OFFSET(_0)?\s+(0x[0-9a-fA-F]+|[0-9]+)$")
38:size_re   = re.compile(r"^#define DT_FLASH_AREA_([0-9A-Z_]+)_SIZE(_0)?\s+(0x[0-9a-fA-F]+|[0-9]+)$")
boot/zephyr/include/target.h
38:    !defined(DT_FLASH_AREA_IMAGE_0_OFFSET) || \
39:    !defined(DT_FLASH_AREA_IMAGE_0_SIZE) || \
40:    !defined(DT_FLASH_AREA_IMAGE_1_OFFSET) || \
41:    !defined(DT_FLASH_AREA_IMAGE_1_SIZE) || \
42:    (!defined(CONFIG_BOOT_SWAP_USING_MOVE) && !defined(DT_FLASH_AREA_IMAGE_SCRATCH_OFFSET)) || \
43:    (!defined(CONFIG_BOOT_SWAP_USING_MOVE) && !defined(DT_FLASH_AREA_IMAGE_SCRATCH_SIZE))
sim/mcuboot-sys/csupport/devicetree.h
12:#define DT_FLASH_AREA_IMAGE_0_ID 1
13:#define DT_FLASH_AREA_IMAGE_1_ID 2
14:#define DT_FLASH_AREA_IMAGE_SCRATCH_ID 3
15:#define DT_FLASH_AREA_IMAGE_2_ID 4
16:#define DT_FLASH_AREA_IMAGE_3_ID 5
boot/zephyr/include/sysflash/sysflash.h
16:                                         DT_FLASH_AREA_IMAGE_0_ID : \
17:                                         DT_FLASH_AREA_IMAGE_0_ID)
19:                                         DT_FLASH_AREA_IMAGE_1_ID : \
20:                                         DT_FLASH_AREA_IMAGE_1_ID)
27:                                         DT_FLASH_AREA_IMAGE_0_ID : \
29:                                         DT_FLASH_AREA_IMAGE_2_ID : \
32:                                         DT_FLASH_AREA_IMAGE_1_ID : \
34:                                         DT_FLASH_AREA_IMAGE_3_ID : \
41:#define FLASH_AREA_IMAGE_SCRATCH    DT_FLASH_AREA_IMAGE_SCRATCH_ID
```

galak replied:
Fixed mcuboot.

mbolivar-nordic commented on subsys/dfu/img_util/flash_img.c:
```diff
@@ -19,9 +19,9 @@
 #include <devicetree.h>
 /* DT_FLASH_AREA_IMAGE_XX_YY values used below are auto-generated by DT */
```
stale

galak replied:
Fixed

mbolivar-nordic commented on subsys/dfu/boot/mcuboot.c:
```diff
@@ -74,15 +74,15 @@ struct mcuboot_v1_raw_header {
 
 /* DT_FLASH_AREA_IMAGE_XX_YY values used below are auto-generated by DT */
```
also stale here

galak replied:
Fixed

mbolivar-nordic reviewed (approved):
Looks overall reasonable as long as the generated IDs are consistent with the legacy IDs and mcuboot is up to date.

galak commented on include/devicetree/fixed-partitions.h:
```diff
@@ -0,0 +1,48 @@
+	IS_ENABLED(DT_COMPAT_fixed_partitions_LABEL_##label##_EXISTS)
+
+/* Given a node id to a fixed partition partition return an
+ * ordinal ID for the partition
```
Updated the comments, let me know if we still need to clarify more.

pabigot reviewed (dismissed):
Bisect shows

> edtlib: handle match of fixed-partition on any bus

causes breakage in `samples/subsys/fs/littlefs` that is reproduced on head when `CONFIG_LEGACY_DEVICETREE_MACROS=y`:
```
In file included from ../../../../../include/devicetree.h:31,
                 from ../../../../../include/arch/arm/aarch32/arch.h:20,
                 from ../../../../../include/arch/cpu.h:19,
                 from ../../../../../include/kernel_includes.h:39,
                 from ../../../../../include/kernel.h:17,
                 from /mnt/nordic/zp/zephyr/arch/arm/core/offsets/offsets_aarch32.c:25,
                 from /mnt/nordic/zp/zephyr/arch/arm/core/offsets/offsets.c:12:
zephyr/include/generated/devicetree_legacy_unfixed.h:2612: warning: "DT_FIXED_PARTITIONS_0_SIZE" redefined
 2612 | #define DT_FIXED_PARTITIONS_0_SIZE                  65536
      | 
zephyr/include/generated/devicetree_legacy_unfixed.h:2239: note: this is the location of the previous definition
 2239 | #define DT_FIXED_PARTITIONS_0_SIZE                  49152
      | 
zephyr/include/generated/devicetree_legacy_unfixed.h:2614: warning: "DT_FIXED_PARTITIONS_0_LABEL" redefined
 2614 | #define DT_FIXED_PARTITIONS_0_LABEL                 "storage"
      | 
zephyr/include/generated/devicetree_legacy_unfixed.h:2241: note: this is the location of the previous definition
 2241 | #define DT_FIXED_PARTITIONS_0_LABEL                 "mcuboot"
      | 
```

because it has:
```
#define DT_FIXED_PARTITIONS_0_BASE_ADDRESS          0x0
#define DT_FIXED_PARTITIONS_0_SIZE                  49152
#define DT_FIXED_PARTITIONS_0_LABEL                 "mcuboot"
#define DT_FIXED_PARTITIONS_0_READ_ONLY             0
#define DT_FIXED_PARTITIONS_0_BASE_ADDRESS          0x0
#define DT_FIXED_PARTITIONS_0_SIZE                  65536
#define DT_FIXED_PARTITIONS_0_LABEL                 "storage"
#define DT_FIXED_PARTITIONS_0_READ_ONLY             0
```
while the original produced only a single setting (the one for `mcuboot`).

galak commented:
> Bisect shows
> 
> > edtlib: handle match of fixed-partition on any bus
> 
> causes breakage in `samples/subsys/fs/littlefs` that is reproduced on head when `CONFIG_LEGACY_DEVICETREE_MACROS=y`:
> 
> ```
> In file included from ../../../../../include/devicetree.h:31,
>                  from ../../../../../include/arch/arm/aarch32/arch.h:20,
>                  from ../../../../../include/arch/cpu.h:19,
>                  from ../../../../../include/kernel_includes.h:39,
>                  from ../../../../../include/kernel.h:17,
>                  from /mnt/nordic/zp/zephyr/arch/arm/core/offsets/offsets_aarch32.c:25,
>                  from /mnt/nordic/zp/zephyr/arch/arm/core/offsets/offsets.c:12:
> zephyr/include/generated/devicetree_legacy_unfixed.h:2612: warning: "DT_FIXED_PARTITIONS_0_SIZE" redefined
>  2612 | #define DT_FIXED_PARTITIONS_0_SIZE                  65536
>       | 
> zephyr/include/generated/devicetree_legacy_unfixed.h:2239: note: this is the location of the previous definition
>  2239 | #define DT_FIXED_PARTITIONS_0_SIZE                  49152
>       | 
> zephyr/include/generated/devicetree_legacy_unfixed.h:2614: warning: "DT_FIXED_PARTITIONS_0_LABEL" redefined
>  2614 | #define DT_FIXED_PARTITIONS_0_LABEL                 "storage"
>       | 
> zephyr/include/generated/devicetree_legacy_unfixed.h:2241: note: this is the location of the previous definition
>  2241 | #define DT_FIXED_PARTITIONS_0_LABEL                 "mcuboot"
>       | 
> ```
> 
> because it has:
> 
> ```
> #define DT_FIXED_PARTITIONS_0_BASE_ADDRESS          0x0
> #define DT_FIXED_PARTITIONS_0_SIZE                  49152
> #define DT_FIXED_PARTITIONS_0_LABEL                 "mcuboot"
> #define DT_FIXED_PARTITIONS_0_READ_ONLY             0
> #define DT_FIXED_PARTITIONS_0_BASE_ADDRESS          0x0
> #define DT_FIXED_PARTITIONS_0_SIZE                  65536
> #define DT_FIXED_PARTITIONS_0_LABEL                 "storage"
> #define DT_FIXED_PARTITIONS_0_READ_ONLY             0
> ```
> 
> while the original produced only a single setting (the one for `mcuboot`).

What should be done here, we fixed a bug w/fixed-partitiions on SPI devices in EDT lib.  This exposes an issue in the old generator.  We can add a feature option to EDT to maintain the old behavior for the old generator.

galak committed:
edtlib: add a flag for old fixed-partitions behavior

allow the old generator to act as it did.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

pabigot commented:
> What should be done here, we fixed a bug w/fixed-partitiions on SPI devices in EDT lib. This exposes an issue in the old generator. We can add a feature option to EDT to maintain the old behavior for the old generator.

I don't know; I'm just  concerned that we won't have tested whether `CONFIG_LEGACY_DEVICETREE_MACROS=y` still works after this and #25009 go in.

galak commented:
> > 
> 
> I don't know; I'm just concerned that we won't have tested whether `CONFIG_LEGACY_DEVICETREE_MACROS=y` still works after this and #25009 go in.

I think that would be the case regardless of CONFIG_LEGACY_DEVICETREE_MACROS setting as we'd mark the old macros as __deprecated otherwise and never have any in tree references/users.  I guess we'd see this particular issue of duplicated defines, but I'm sure there are other issues we'd miss.

galak committed:
gen_defines: Rework implementation of write_child_functions

Rework how write_child_functions to match how we do the code for
DT_FOREACH_OKAY_INST.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

mbolivar-nordic commented:
Test results:

assemble.py runs without error (cc @brix):

```
$ ./bootloader/mcuboot/scripts/assemble.py -b build_mcuboot/ -p build/zephyr/zephyr.signed.bin -s build/zephyr/zephyr.signed.bin -o /tmp/out
partition MCUBOOT, pos=0, offset=0
partition IMAGE_0, pos=37592, offset=49152
partition IMAGE_1, pos=229972, offset=471040
```

smp_svr: I can build and mostly flash with the instructions online in the Zephyr docs (which use west instead of assemble.py), booting the application via mcuboot.

I could not run the DFU upload step, but I think that's because my laptop's Bluetooth stopped working. I see some HCI errors in `dmesg` now and I can't get `list` to work in `bluetoothctl`. So I don't consider this a real failure.

I believe the legacy and new flash area IDs are consistent with one another:

```
$ grep -e PARTITION_ID -e FLASH_AREA_[0-9].*_ID devicetree_*unfixed.h
devicetree_legacy_unfixed.h:#define DT_FLASH_AREA_0_ID                          0
devicetree_legacy_unfixed.h:#define DT_FLASH_AREA_1_ID                          1
devicetree_legacy_unfixed.h:#define DT_FLASH_AREA_2_ID                          2
devicetree_legacy_unfixed.h:#define DT_FLASH_AREA_3_ID                          3
devicetree_legacy_unfixed.h:#define DT_FLASH_AREA_4_ID                          4
devicetree_unfixed.h:#define DT_N_S_soc_S_flash_controller_4001e000_S_flash_0_S_partitions_S_partition_0_PARTITION_ID 0
devicetree_unfixed.h:#define DT_N_S_soc_S_flash_controller_4001e000_S_flash_0_S_partitions_S_partition_c000_PARTITION_ID 1
devicetree_unfixed.h:#define DT_N_S_soc_S_flash_controller_4001e000_S_flash_0_S_partitions_S_partition_73000_PARTITION_ID 2
devicetree_unfixed.h:#define DT_N_S_soc_S_flash_controller_4001e000_S_flash_0_S_partitions_S_partition_da000_PARTITION_ID 3
devicetree_unfixed.h:#define DT_N_S_soc_S_flash_controller_4001e000_S_flash_0_S_partitions_S_partition_f8000_PARTITION_ID 4
```

I can't find anything wrong with this.

galak commented:
> > What should be done here, we fixed a bug w/fixed-partitiions on SPI devices in EDT lib. This exposes an issue in the old generator. We can add a feature option to EDT to maintain the old behavior for the old generator.
> 
> I don't know; I'm just concerned that we won't have tested whether `CONFIG_LEGACY_DEVICETREE_MACROS=y` still works after this and #25009 go in.

Fair, but isn't that true of any deprecated functionality in Zephyr.  Since any tests would cause warnings they would never pass sanitycheck.

pabigot commented:
> > I don't know; I'm just concerned that we won't have tested whether `CONFIG_LEGACY_DEVICETREE_MACROS=y` still works after this and #25009 go in.
> 
> Fair, but isn't that true of any deprecated functionality in Zephyr. Since any tests would cause warnings they would never pass sanitycheck.

I'm not talking about sanitycheck, I'm talking about peoples out-of-tree applications where they aren't quite ready to switch to the non-legacy macros.  If there were no expectation that `CONFIG_LEGACY_DEVICETREE_MACROS=y` would work in that case, why would we even keep the option?

galak commented:
> > > I don't know; I'm just concerned that we won't have tested whether `CONFIG_LEGACY_DEVICETREE_MACROS=y` still works after this and #25009 go in.
> > 
> > 
> > Fair, but isn't that true of any deprecated functionality in Zephyr. Since any tests would cause warnings they would never pass sanitycheck.
> 
> I'm not talking about sanitycheck, I'm talking about people's out-of-tree applications where they aren't quite ready to switch to the non-legacy macros. If there were no expectation that `CONFIG_LEGACY_DEVICETREE_MACROS=y` would work in that case, why would we even keep the option?

The option is meant as the means to "deprecate" the functionality vs marking every define as __deprecated which has some issues if a defined is used in an #ifdef clause.

I'm not disagreeing with your point, I'm just saying we have this issue more broadly than just here for deprecated functionality.

galak commented:
> > > > I don't know; I'm just concerned that we won't have tested whether `CONFIG_LEGACY_DEVICETREE_MACROS=y` still works after this and #25009 go in.
> > > 
> > > 
> > > Fair, but isn't that true of any deprecated functionality in Zephyr. Since any tests would cause warnings they would never pass sanitycheck.
> > 
> > 
> > I'm not talking about sanitycheck, I'm talking about people's out-of-tree applications where they aren't quite ready to switch to the non-legacy macros. If there were no expectation that `CONFIG_LEGACY_DEVICETREE_MACROS=y` would work in that case, why would we even keep the option?
> 
> The option is meant as the means to "deprecate" the functionality vs marking every define as __deprecated which has some issues if a defined is used in an #ifdef clause.
> 
> I'm not disagreeing with your point, I'm just saying we have this issue more broadly than just here for deprecated functionality.

We probably should have some tests/deprecated/FOO or something that handles maintaining to the the deprecated functionality for the period of time that its still in tree.

galak committed:
tests: deprecated: dts: Test deprecated DTS functionality

This this covers two small aspects of DTS functionality: first ensuring
the legacy generation script continues to generally function and second
that FLASH_AREA id's are the same between the old and new generator.

This test will be removed when the legacy generator is removed.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak commented:
@pabigot I added a tests/deprecated/dts that at least enables `CONFIG_LEGACY_DEVICETREE_MACROS=y` and does some simple flash map tests.

galak commented:
Is based on needs #25009

mbolivar-nordic reviewed (approved):
This looks to me like it gets the job done of deprecating the legacy devicetree macros for v2.3, which @carlescufi has said is a priority (not shipping with any deprecated functionality enabled, that is). I also don't see anything obvious that it breaks in terms of legacy support.

I think getting @nvlsianpu 's review would be good before merging, though, since he knows this area much better than I do.

As discussed on dev-review last week and Slack this morning, it seems like a few TSC members agree that extending the release freeze by a day or two due to late-breaking DT changes would not be the end of the world.

nvlsianpu reviewed (commented):
this umbrella PR looks good, although can't accept mcuboot path without deeper analyze I can't do this weekend. Will do as very first thing on Monday.

pabigot commented on scripts/dts/edtlib.py:
```diff
@@ -175,13 +176,19 @@ def __init__(self, dts, bindings_dirs, warn_file=None,
 
         self._warn_reg_unit_address_mismatch = warn_reg_unit_address_mismatch
         self._default_prop_types = default_prop_types
+        self._fixed_partions_no_bus = support_fixed_partitions_on_any_bus
```
Please rename the property to either `_fixed_partitions_no_bus` or `_fixed_partitions_on_bus`, whichever is correct.  Either way `partions` isn't right.

galak replied:
Fixed.

carlescufi commented:
@galak @mbolivar-nordic 
This can be considered a fix since the majority of the API changes have been introduced already, so I propose we merge this after the `-rc1` tag, given that @nvlsianpu cannot review before Monday.

pabigot commented on include/devicetree/fixed-partitions.h:
```diff
@@ -0,0 +1,57 @@
+ */
+
+/**
+ * @brief Get fixed partition that matches label
```
I would like the documentation to be more clear that this is the `label` property in the node, not a node label attached to the node.

Perhaps: 
* `@brief Get fixed partition with "label" property that matches label`
* `@param label lowercase-and-underscores value for label property`

Applies to both macros in this patch.

galak replied:
Update the comments.

pabigot commented on doc/guides/dts/macros.bnf:
```diff
@@ -148,7 +150,8 @@ other-macro =/ %s"DT_CHOSEN_" dt-name
 ;   #define DT_COMPAT_HAS_OKAY_vnd_dev 1
 other-macro =/ %s"DT_COMPAT_HAS_OKAY_" dt-name
-
+; Allows mapping a label to a device node
```
Make this `Allows mapping a "label" property to a compatible node.`  Or something that makes clear the label is a property.

galak replied:
Done

pabigot reviewed (changes requested):
Some documentation clarifications and one code change.

pabigot reviewed (approved):
Tested with mcuboot and `samples/subsys/mgmt/mcumgr/smp_svr` with #25167 which included filesystem support on an external SPI NOR flash (`storage` moved to external flash in a local overlay not part of that PR).

hakonfam commented on tests/deprecated/dts/CMakeLists.txt:
```diff
@@ -0,0 +1,8 @@
+cmake_minimum_required(VERSION 3.13.1)
+
+find_package(Zephyr HINTS $ENV{ZEPHYR_BASE})
+project(hello_world)
```
Change this? Also the commit message for the last commit seems to need some fixing up.

```
This this covers to small aspects of DTS functionality, one that the
legacy generation script continues to generally function and that
FLASH_AREA id's are the same between the old and new generator.
```

hakonfam reviewed (commented):
just nits, the last three commit messages needs to be double checked

hakonfam commented on include/storage/flash_map.h:
```diff
@@ -209,6 +209,18 @@ int flash_area_has_driver(const struct flash_area *fa);
  */
 struct device *flash_area_get_device(const struct flash_area *fa);
 
+#define FLASH_AREA_LABEL_EXISTS(label) \
+	DT_HAS_FIXED_PARTITION_LABEL(label)
+
+#define FLASH_AREA_ID(label) \
+	DT_FIXED_PARTITION_ID(DT_NODE_BY_FIXED_PARTITION_LABEL(label))
+
+#define FLASH_AREA_OFFSET(label) \
+	DT_REG_ADDR(DT_NODE_BY_FIXED_PARTITION_LABEL(label))
+
+#define FLASH_AREA_SIZE(label) \
+	DT_REG_SIZE(DT_NODE_BY_FIXED_PARTITION_LABEL(label))
+
```
We should allow users to provide their own definition of these symbols. This could be the default definition included, but (like with `flash_map_default.c`) it should be possible to override this through a kconfig value.

Currently any dowstream with their own partition definitions is required to maintain a patch of this file.

pabigot replied:
My initial thoughts for solving this were to in the future extend this macro to take a second parameter identifying the provider of partition information.  The default behavior would be as if that parameter were `fixed_partition` which would through magic pass to the current solution, but if it were say `partition_mgr` would route to a different macro or possibly ultimately a function.

The thought is that we do need label uniqueness within a given provider of flash region managers.  Attempting to enforce it across all region managers isn't reasonable IMO, and limiting applications to a single region manager (devicetree or other) is also not reasonable.

A way to resolve the conflict is to pair the label with a provider, using the existing devicetree provider as the default.

nvlsianpu replied:
This was consulted with me. @mbolivar This is the simple way we can inject our partitions info.

hakonfam replied:
OK, then there must be something like a kconfig option passed in as the second argument to avoid having to patch the file.

I'm fine with this as long as it allows us to not need patches in any of the upstream source files.

nvlsianpu commented on west.yml:
```diff
@@ -86,10 +86,10 @@ manifest:
       path: modules/crypto/mbedtls
     - name: mcuboot
-      revision: 5657d00e662adbd32addc8525862249b631334c5
+      revision: pull/17/head
```
upgrade sha
